volume/set.go

```diff
@@ -0,0 +1,169 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package volume
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/continuity/fs"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+const (
+	vmVolumePath = "/vmv"
+	fsType       = "ext4"
+)
+
+// Set is a set of volumes.
+type Set struct {
+	volumes map[string]*Volume
+	tempDir string
+}
+
+// NewSet returns a new volume set.
+func NewSet() *Set {
+	return NewSetWithTempDir(os.TempDir())
+}
+
+// NewSetWithTempDir returns a new volume set and creates all temporary files under the tempDir.
+func NewSetWithTempDir(tempDir string) *Set {
+	return &Set{volumes: make(map[string]*Volume), tempDir: tempDir}
+}
+
+// Add a volume to the set.
+func (vs *Set) Add(v *Volume) {
+	vs.volumes[v.name] = v
+}
+
+func (vs *Set) createDiskImage(ctx context.Context, size int64) (string, error) {
+	var err error
+
+	f, err := os.CreateTemp(vs.tempDir, "")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+			os.Remove(f.Name())
+		}
+	}()
+
+	err = f.Truncate(size)
+	if err != nil {
+		return "", err
+	}
+
+	err = f.Close()
+	if err != nil {
+		return "", err
+	}
+
+	out, err := exec.CommandContext(ctx, "mkfs."+fsType, "-F", f.Name()).CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("failed to execute mkfs.%s: %s: %w", fsType, out, err)
+	}
+
+	return f.Name(), nil
+}
+
+func mountDiskImage(source, target string) error {
+	return mount.All([]mount.Mount{{Type: fsType, Source: source, Options: []string{"loop"}}}, target)
+}
+
+// PrepareDriveMount returns a FirecrackerDriveMount that could be used with CreateVM.
+func (vs *Set) PrepareDriveMount(ctx context.Context, size int64) (*proto.FirecrackerDriveMount, error) {
+	path, err := vs.createDiskImage(ctx, size)
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err := os.MkdirTemp("", "")
```

Comment:
`dir` never gets removed. Also, if `mount.Unmount` fails, the error is lost and the directory can't be unmounted since there's no reference to `dir` anymore. I'd like to see some more robust error handling and some place that the client can either get information to do manual cleanup, or trigger a retry, or connect this to GC.

Reply:
You don't have to approve a PR if you find critical issues :)

Let me fix...

Reply:
Okay. The last revision shouldn't have any resource leaks.

It is probably difficult to connect to containerd's GC currently, since these Go code is basically living outside of containerd.